import grpc
import banque_pb2
import banque_pb2_grpc

def main():
    print("ğŸš€ Test du service gRPC Banque")
    print("Connexion au serveur sur localhost:9090...")
    
    # Connexion au serveur Java
    channel = grpc.insecure_channel('localhost:9090')
    stub = banque_pb2_grpc.BanqueServiceStub(channel)
    
    try:
        # Test 1: CrÃ©er un client
        print("\n1. CrÃ©ation d'un client...")
        reponse_client = stub.CreerClient(
            banque_pb2.ClientRequest(
                numero_tel="+212612345678",
                solde_initial=100000
            )
        )
        print(f"âœ… Client crÃ©Ã©: {reponse_client.message}")
        print(f"   ğŸ“± NumÃ©ro: {reponse_client.numero_tel}")
        print(f"   ğŸ’° Solde: {reponse_client.solde_actuel}")
        
        # Test 2: Consulter le solde
        print("\n2. Consultation du solde...")
        reponse_solde = stub.ConsulterSolde(
            banque_pb2.SoldeRequest(numero_tel="+212612345678")
        )
        print(f"âœ… Solde consultÃ©: {reponse_solde.solde}")
        print(f"   ğŸ“… Date: {reponse_solde.date_consultation}")
        
        # Test 3: Faire un transfert
        print("\n3. Transfert entre clients...")
        reponse_transfert = stub.EffectuerTransfert(
            banque_pb2.TransfertRequest(
                emetteur="+212612345678",
                destinataire="+212698765432", 
                montant=5000
            )
        )
        print(f"âœ… Transfert: {reponse_transfert.message}")
        print(f"   ğŸ“‹ RÃ©fÃ©rence: {reponse_transfert.reference}")
        print(f"   ğŸ’° Nouveau solde: {reponse_transfert.nouveau_solde_emetteur}")
        
        # Test 4: Voir les transactions (streaming)
        print("\n4. Streaming des transactions...")
        print("   RÃ©cupÃ©ration de 3 transactions...")
        reponse_stream = stub.StreamTransactions(
            banque_pb2.StreamRequest(
                numero_tel="+212612345678",
                max_transactions=3
            )
        )
        
        for i, transaction in enumerate(reponse_stream):
            print(f"   ğŸ“ˆ Transaction {i+1}:")
            print(f"      ğŸ†” ID: {transaction.id}")
            print(f"      ğŸ“ Type: {transaction.type}")
            print(f"      ğŸ’° Montant: {transaction.montant}")
            print(f"      ğŸ“… Date: {transaction.date}")
            print(f"      ğŸ“‹ Description: {transaction.description}")
        
        print("\nğŸ‰ Tous les tests sont rÃ©ussis!")
        
    except Exception as e:
        print(f"âŒ Erreur: {e}")

if __name__ == '__main__':
    main()